hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemLease.java

```diff
@@ -0,0 +1,249 @@
+    fs.getAbfsStore().getAbfsConfiguration()
+        .setAzureSingleWriterDirs(testFilePath.getParent().toString());
+    fs.getAbfsStore().updateSingleWriterDirs();
+
+    FSDataOutputStream out;
+    out = fs.create(testFilePath);
+    out.write(0);
+    Assert.assertTrue("Output stream should have lease",
+        ((AbfsOutputStream) out.getWrappedStream()).hasLease());
+    out.close();
+    Assert.assertFalse("Output stream should not have lease after close",
+        ((AbfsOutputStream) out.getWrappedStream()).hasLease());
+    Assert.assertTrue(fs.getAbfsStore().areLeasesFreed());
+  }
+
+  @Test(timeout = TEST_EXECUTION_TIMEOUT)
+  public void testWriteAfterBreakLease() throws IOException {
+    final AzureBlobFileSystem fs = getFileSystem();
+    final Path testFilePath = new Path(path(methodName.getMethodName()), TEST_FILE);
+    fs.mkdirs(testFilePath.getParent());
+    fs.getAbfsStore().getAbfsConfiguration()
+        .setAzureSingleWriterDirs(testFilePath.getParent().toString());
+    fs.getAbfsStore().updateSingleWriterDirs();
+
+    FSDataOutputStream out;
+    out = fs.create(testFilePath);
+    out.write(0);
+    out.hsync();
+
+    fs.breakLease(testFilePath);
+    try {
+      out.write(1);
+      out.hsync();
+      Assert.fail("Expected exception on write after lease break");
+    } catch (IOException e) {
+      Assert.assertTrue("Unexpected error message: " + e.getMessage(),
+          e.getMessage().contains(ERR_LEASE_EXPIRED));
+    }
+    try {
+      out.close();
+      Assert.fail("Expected exception on close after lease break");
+    } catch (IOException e) {
+      Assert.assertTrue("Unexpected error message: " + e.getMessage(),
+          e.getMessage().contains(ERR_LEASE_EXPIRED));
+    }
+
+    Assert.assertTrue(((AbfsOutputStream) out.getWrappedStream()).isLeaseFreed());
+
+    try (FSDataOutputStream out2 = fs.append(testFilePath)) {
+      out2.write(2);
+      out2.hsync();
+    }
+
+    Assert.assertTrue(fs.getAbfsStore().areLeasesFreed());
+  }
+
+  @Test(timeout = LONG_TEST_EXECUTION_TIMEOUT)
+  public void testLeaseFreedAfterBreak() throws IOException {
+    final AzureBlobFileSystem fs = getFileSystem();
+    final Path testFilePath = new Path(path(methodName.getMethodName()), TEST_FILE);
+    fs.mkdirs(testFilePath.getParent());
+    fs.getAbfsStore().getAbfsConfiguration()
+        .setAzureSingleWriterDirs(testFilePath.getParent().toString());
+    fs.getAbfsStore().updateSingleWriterDirs();
+
+    FSDataOutputStream out = null;
+    try {
+      out = fs.create(testFilePath);
+      out.write(0);
+
+      fs.breakLease(testFilePath);
+      while (!((AbfsOutputStream) out.getWrappedStream()).isLeaseFreed()) {
+        try {
+          Thread.sleep(1000);
+        } catch (InterruptedException e) {
+        }
+      }
+    } finally {
+      try {
+        if (out != null) {
+          out.close();
+        }
+        // exception might or might not occur
+      } catch (IOException e) {
```

Comment:
GenericTestUtils lets you assert something is in the error message. Its critical to rethrow (maybe wrapped) the exception if it is not the one you were expecting